src/include/simeng/config/SimInfo.hh

```diff
@@ -0,0 +1,218 @@
+   * option. */
+  template <typename T>
+  static T getValue(ryml::ConstNodeRef node) {
+    T val;
+    node >> val;
+    return val;
+  }
+
+  /** A getter function to retrieve the config file path. */
+  static std::string getConfigPath() { return getInstance()->configFilePath_; }
+
+  /** A setter function to set the config file path. */
+  static void setConfigPath(std::string path) {
+    getInstance()->configFilePath_ = path;
+  }
+
+  /** A getter function to retrieve the simulation mode of the current SimEng
+   * instance. */
+  static SimulationMode getSimMode() { return getInstance()->mode_; }
+
+  /** A getter function to retrieve the simulation mode of the current SimEng
+   * instance as a string. */
+  static std::string getSimModeStr() { return getInstance()->modeStr_; }
+
+  /** A getter function to retrieve which ISA the current simulation is using.
+   */
+  static ISA getISA() { return getInstance()->isa_; }
+
+  /** A getter function to retrieve a vector of {size, number} pairs describing
+   * the available architectural registers. */
+  static const std::vector<simeng::RegisterFileStructure>& getArchRegStruct() {
+    return getInstance()->archInfo_->getArchRegStruct();
+  }
+
+  /** A getter function to retrieve a vector of {size, number} pairs describing
+   * the available physical registers. */
+  static const std::vector<simeng::RegisterFileStructure>& getPhysRegStruct() {
+    return getInstance()->archInfo_->getPhysRegStruct();
+  }
+
+  /** A getter function to retrieve a vector of uint16_t values describing
+   * the quantities of physical registers available. */
+  static const std::vector<uint16_t>& getPhysRegQuantities() {
+    return getInstance()->archInfo_->getPhysRegQuantities();
+  }
+
+  /** A getter function to retrieve a vector of Capstone sysreg enums for
+   * all the system registers that should be utilised in simulation. */
+  static const std::vector<uint64_t>& getSysRegVec() {
+    return getInstance()->archInfo_->getSysRegEnums();
+  }
+
+  /** A getter function to retrieve whether or not the special files
+   * directories should be generated. */
+  static const bool getGenSpecFiles() {
+    return getInstance()->genSpecialFiles_;
+  }
+
+  /** A utility function to rebuild/construct member variables/classes. For use
+   * if the configuration used changes during simulation (e.g. during the
+   * execution of a test suite). */
+  static void reBuild() { getInstance()->extractValues(); }
+
+ private:
+  SimInfo() {
+    // Set the validated config file to be the current default config
+    // generated by the default constructor of ModelConfig
+    validatedConfig_ = modelConfig_.getConfig();
+    extractValues();
+  }
+
+  /** Gets the static instance of the SimInfo class. */
+  static std::unique_ptr<SimInfo>& getInstance() {
+    static std::unique_ptr<SimInfo> SimInfoClass = nullptr;
+    if (SimInfoClass == nullptr) {
+      SimInfoClass = std::unique_ptr<SimInfo>(new SimInfo());
+    }
```

Comment:
Why is there an if statement here -- won't it always have the value of 'nullptr' as we've set it to that in the line above?

Reply:
As we initialise a static unique_ptr to a new SimInfo object, the variable will retian its state between all function calls. So, on the first call to the function `SimInfoClass` will indeed be nullptr, but on all subsequent calls the same `SimInfoClass` vaiable has already been instantiated.

Reply:
gotcha -- makes sense